### ISSUE TITLE
[feedback request] domvm experience

### ISSUE THREAD

**Issue by leeoniya** [ludlovian/civic-society-members#1]
Hey @ludlovian ,

If you ever get a chance, it would be great to get some feedback (both positive and negative) about your experience with domvm (in general and vs Mithril) from a DX and performance perspective (mobile would probably show largest gains). I'm always curious how others' large-ish apps work out. A gauge of your typical & max dom node count would also be helpful as a reference point, too (just run `document.querySelectorAll("*").length` in the console).

thanks!

**Comment by ludlovian** [comment]
Hi Leon,

Really pleased with domvm. My app is really a mickey mouse hobby thing (for a local charity), as I'm too old and too long in the tooth to do real work anymore 😉, but at 3.5k dom elements is reasonably sized.

I wrote it a few years ago in Bootstrap/React and wanted to refresh it a few months ago, so the natural progression was Material/Preact. But I still didn't like the inherent "magic" of React, which Preact has to follow. So I searched for alternatives, finding Mithril soon after which seemed much easier to deal with. But it still had magic redraws, and quite a lot of boilerplate for stateful components. So, I rewrote it in Mithril, but it still didn't feel "right" to me. I briefly considered writing my own vdom just as a technical challenge, and certainly thought through how it would be architected (but was too lazy to do anything further)

I came across domvm in a Mithril thread, and took a look. It seemed exactly how I would have written a vdom engine - comparing new templates against existing vnodes and only applying changes to the DOM, rather than comparing with the DOM each time (like React/Preact does). Not over optimising cheap things (creating vnodes), but carefully minimising expensive things (interaction with DOM). And avoiding the temptation to do too much, instead sticking to one thing and doing it well.

I'm also quite a fan of opinionated libraries ("do it this way, not that way"), as they tell me the best way to work with the library. Mitrhil's everything-under-the-sun approach to components (ES6 classes, closures, plain objects) just felt too open-ended ("just tell me which way I should do it already"), but they have to support an installed base that uses it all those different ways, so...

Most of my needs for "components/views" are to hold local state and do cleanup (e.g. to set up & tear down Material Components), and domvm allows me to do that on vnodes (didInsert, willRecycle, willRemove) without needing components. I only now use views for UI elements that need to decide themselves when to rerender.  Stateful elements are all created using template creating functions. 30-line Material component wrappers in Preact / Mithril become 10-line pure functions in domvm.

I'm a coder at heart, and don't like to use a library unless I've read the source code.

- React is a nightmare to read, so large you cannot really get your hands around it. But then, it is the market leader, on its 16th major version, and it shows in the code base.
- Mithrill seems quite confused in its structure - probably because of its venerable age and the batteries included approach, so I found myself getting lost very often (e.g. vnode.state is *not* the vnode's internal state but seems to have everything packed into it)
- Preact is a great intellectual challenge to read, having been code-golfed to save that last extra byte, but not so easy if you want to work out why that text box is rendering the way you think it should. Everyone should read Preact's code base just to learn the limits of the language!
- domvm, however, was very easy to read and follow - which is essential because the code is pretty much the best documentation for how it works.

If I had a suggestion, it would be around tips for avoiding views/components and using template-generating functions. In particular:

- use didInsert to store initial state on the vnode (under node.data? node.state?)
- on re-render, any custom data has to be copied across from old vnode to new one n the willRecycle hook
- teardown can be placed on the willRemove hook
- and that's it - unless you need to control your own redraws, most things that would be components in React can be (and should be) turned into template generating functions.

In summary, domvm really chimed with my desire to be close the metal - one step up from jQuery or VanillaJS - keeping the advantages of React's declarative top-down approach where it works, but without layering too much into the library.

Thanks for your work on this,

Alan

**Comment by leeoniya** [comment]
Wow, very thorough and interesting, thanks!

> If I had a suggestion, it would be around tips for avoiding views/components and using template-generating functions.

yeah this might be useful to add as an alternative to Views. though with Views you avoid the need for the setup and recycle hooks. i'm a bit hesitant to recommend an over-avoidance of Views in favor of this more verbose vnode wiring, but i'll give it some thought for sure.

> * use `didInsert` to store initial state on the vnode (under node.data? node.state?)

for intializing stuff on the vnodes, i'd recommend using `willInsert`, since `did*` hooks have to fire later from a queue after a forced browser repaint. if the didQueue is empty after a redraw call, no forced repaint is performed [1].

> * on re-render, any custom data has to be copied across from old vnode to new one the willRecycle hook

hmm, this isn't the first time someone's needed to do this. your vnodes are keyed, which is good, so you don't accidentally get mismatched sibling vnodes getting cross-recycled.

> my desire to be close the metal

i always cringe a bit at this terminology when talking about JS & DOM, but sure :)

> Thanks for your work on this,

I do what i can, glad to see it getting some use here and there.

Out of curiosity, have you done any perf testing or do you leave that for only when things feel slow? Often you can be wasting a lot of cpu cycles/battery (on a fast machine) long before things actually begin to feel slow in the UI. For one-shot interactions (like click or touch), you wont feel it until it's about ~100ms; for drag/scroll, more like 3-5ms. i usually make sure that my redraws are within 16ms. With 3.5k nodes you should be okay (even regenerating everything), but that's about as high as i care to go myself. This is a reason why i use my own flexbox grid rather than bootstrap, so i dont have to have so many wrapping elements (each col can also be a grid and the cell spacing isnt faked with padding).

[1] https://github.com/domvm/domvm/blob/50ad5bbdecb12dd73938e9d0e2e64abe57dedb30/src/view/hooks.js#L28-L36

**Comment by ludlovian** [comment]
> I'm a bit hesitant to recommend an over-avoidance of Views in favor of this more verbose vnode wiring, but i'll give it some thought for sure.

My own opinionated view is that components/views are used for three reasons - in descending order of frequency:

1. Encapsulation of a piece of code to add structure and reduce complexity
2. Provision of local state data and setup/teardown hooks
3. Independent entity that can control its own re-rendering

In the React world, you use components for all three (albeit  case 1 can be a pure functional one), so people get used to using components for everything. Hence you have components that render components that render ..... and so on.

In the domvm world, views are used for case 3, can be used for case 2, and are not really needed for the most common case of simple encapsulation, where a plain old function will suffice. People coming from React-y backgrounds may not realise this (I didn't) and start writing views everywhere, when they are really not needed. _Your docs already explain this, but I ignored it. Like most people, I imagine I know better...until I realise I don't._

> for intializing stuff on the vnodes, i'd recommend using `willInsert`

I'm attaching Material handlers to actual DOM elements, so I'd rather wait until they've actually been rendered as that's what Material expects, but I agree with your general point.

> > my desire to be close the metal
>
> i always cringe a bit at this terminology when talking about JS & DOM, but sure :)

I agree. Should have said closer to the DOM. Which is still a loooong way from the metal!

React's big breakthrough was allowing UI to be encoded in a declarative way. Its drawback is its dogmatic insistence that everything is declarative, which creates an inexorable pressure to move state upwards until you have an army of simple components controlled by a mass of complex logic at the top. This in turn requires more solutions to deal with it (e.g. Redux / global state management).

In the real world of humans, they expect some imperative actions (I click this down here and that happens). It is often easier to code this as imperative logic, whilst keeping the declarative approach for how it looks.

domvm's approach of being close to the DOM allows you to mix the two styles with ease. One step up from VanillaJS, with all the ease of declarative layouts, but with the  ability for views to expose APIs so that they can react (no pun intended) to imperative commands.

> Out of curiosity, have you done any perf testing or do you leave that for only when things feel slow?

Sadly, no. I have the luxury of elderly users with speedy computers and low expectations. I rarely test my apps. Having been in senior management, I know the value of rigorous testing, but I am too lazy to do it. Hypocrisy is one of the great perks of senior management.

But if I was worried about performance, I wouldn't be using Material components. Like supermodels, they are pretty, but slow. But that doesn't mean *I* haven't noticed a speedup. Mithril & domvm feel much snappier than the equivalent React/Preact code. The logic of recreating everything (or most things in domvm's case) on each render seems sooo much faster, and will only improve as Moore's law bites. React's approach was a premature optimisation of its time, which has now been fossilised into its architecture 😮.

Alan

**Comment by leeoniya** [comment]
@ludlovian sorry for the prolonged radio silence (real life responsibilities and all that). thanks for your feedback!
